### PR TITLE
[TESTMERGE CANDIDATE] [SEMI-MODULAR] AI and hologram QoL stuff + misc bugfixes

### PR DESCRIPTION
**Pull request by Useroth**
:cl:
refactor: Standarized flavourtext displaying for all silicons (AIs, borgs, pAI holo forms)
add: AI eye can now see emotes (or to be specific, **non-spoken** messages emitted from things)
add: It's now possible to examine holopad holograms during the call to actually examine the character represented by the holo
add: It's now possible to properly send emotes, subtles and subtlers through holopad holograms
add: It's now possible to properly see emotes, subtles and subtlers through holopad holograms
add: BONUS because I hate snowflake code: the above should work with regular character-made holocalls too
fix: SQL errors when using subtle or subtler without db connection
/:cl:

I hate how much much spaghetti I had to wade through to do this semi-properly.

tl;dr
image

also, probably merge after the freeze is done with, it can be TMed for now

### CODE REVIEW

File: code/datums/emotes.dm [hunk 1]
```diff
@@ -137,6 +131,18 @@
...
+	if(h)
+		if(emote_type & (EMOTE_AUDIBLE | EMOTE_VISIBLE))
+			h.audible_message(msg, deaf_message = "<span class='emote'>You see how <b>[user]</b> [msg]</span>", audible_message_flags = EMOTE_MESSAGE)
+		else if(emote_type & EMOTE_VISIBLE)
+			h.visible_message(msg, visible_message_flags = EMOTE_MESSAGE)
+		if(emote_type & EMOTE_IMPORTANT)
+			for(var/mob/living/viewer in viewers(world.view, h))
+				if(viewer.is_blind() && !viewer.can_hear())
+					to_chat(viewer, msg)
+	// SKYRAT EDIT -- END
```
Comment (GoldenAlpharex) [inline code/datums/emotes.dm]: Single-letter variables are bad, give `h` a better name.

Also use `span_emote()` instead of baking the span classes in the string.

File: modular_skyrat/modules/altborgs/code/robot_defines.dm [hunk 2]
```diff
@@ -1,4 +1,6 @@
+/mob/living/silicon
+	var/datum/examine_panel/tgui = new() //create the datum
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/altborgs/code/robot_defines.dm]: Please rename the `tgui` variables to something that means something without seeing the definition, like `examine_panel` (that would work).

File: modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/mob/living/silicon/proc/get_silicon_flavortext()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm:5]: Document this proc please.

File: modular_skyrat/modules/verbs/code/subtle.dm [hunk 4]
```diff
@@ -45,6 +45,14 @@
 
 	var/list/viewers = get_hearers_in_view(SUBTLE_DEFAULT_DISTANCE, user)
 
+	var/obj/effect/overlay/holo_pad_hologram/h = GLOB.hologram_impersonators[user]
+	if(h)
+		viewers |= get_hearers_in_view(SUBTLE_DEFAULT_DISTANCE, h)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/verbs/code/subtle.dm]: Use something that's different from `h` here too, please.

File: modular_skyrat/modules/verbs/code/subtle.dm [hunk 5]
```diff
@@ -83,9 +91,17 @@
 			return FALSE
 
 		var/list/in_view = get_hearers_in_view(1, user)
+
+		var/obj/effect/overlay/holo_pad_hologram/h = GLOB.hologram_impersonators[user]
+		if(h)
+			in_view |= get_hearers_in_view(1, h)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/verbs/code/subtle.dm]: Rename `h` here as well please.

File: modular_skyrat/modules/verbs/code/subtle.dm [hunk 6]
```diff
@@ -113,14 +129,28 @@
 
 	subtler_message = span_emote("<b>[user]</b>[space]<i>[user.say_emphasis(subtler_message)]</i>")
 
-	if(istype(target))
+	if(istype(target, /mob))
+		var/mob/M = target
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/verbs/code/subtle.dm]: Why did you do this? It was already checking that `target` was indeed of the type it was declared as, which is `/mob` here.

Plus now you're using a single-letter var, which is definitely no bueno.

File: modular_skyrat/modules/verbs/code/subtle.dm [hunk 7]
```diff
@@ -113,14 +129,28 @@
...
+	else if(istype(target, /obj/effect/overlay/holo_pad_hologram))
+		var/obj/effect/overlay/holo_pad_hologram/holo = target
+		if(holo.Impersonation?.client)
+			holo.Impersonation.show_message(subtler_message, alt_msg = subtler_message)
 	else
 		var/ghostless = get_hearers_in_view(target, user) - GLOB.dead_mob_list
+
+		var/obj/effect/overlay/holo_pad_hologram/h = GLOB.hologram_impersonators[user]
+		if(h)
+			ghostless |= get_hearers_in_view(target, h)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/verbs/code/subtle.dm]: single letter var bad, rename `h`.

File: modular_skyrat/modules/customization/modules/mob/living/silicon/topic.dm [hunk 8]
```diff
@@ -0,0 +1,5 @@
+/mob/living/silicon/Topic(href, href_list)
+	. = ..()
+	if(href_list["lookup_info"] == "open_examine_panel")
+		tgui.holder = src
+		tgui.ui_interact(usr) //datum has a tgui component, here we open the window
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/mob/living/silicon/topic.dm]: Don't forget to rename `tgui` (that I mentioned in an earlier comment).

File: modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+	flavor_text_link = span_notice("[silicon_preview_text]... <a href='?src=[REF(src)];lookup_info=open_examine_panel'>Look closer?</a>")
+
+	if (flavor_text_link)
+		. += flavor_text_link
+
+	if(client)
+		var/erp_status_pref = client.prefs.read_preference(/datum/preference/choiced/erp_status)
+		if(erp_status_pref && erp_status_pref != "disabled")
+			. += span_notice("ERP STATUS: [erp_status_pref]")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm:19]: Do we really always display the `ERP STATUS:` underneath that string all the time?

File: modular_skyrat/master_files/code/game/machinery/hologram.dm [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+	if(h.Impersonation)
+		GLOB.hologram_impersonators[user] = h
+		h.become_hearing_sensitive() // Well, we need to show up on "get_hearers_in_view()"
+	. = ..()
+
+/obj/machinery/holopad/clear_holo(mob/living/user)
+	var/obj/effect/overlay/holo_pad_hologram/h = GLOB.hologram_impersonators[user]
+	if(h)
+		h.lose_hearing_sensitivity()
+		GLOB.hologram_impersonators -= user
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/game/machinery/hologram.dm]: Don't use `h`, please. Better name for this.

Also, if possible, rename the `Impersonation` variable to `impersonation` too, if that's a Skyrat variable and not a /tg/ one.

File: code/modules/pai/pai.dm [hunk 11]
```diff
@@ -169,6 +169,9 @@
 /mob/living/silicon/pai/examine(mob/user)
 	. = ..()
 	. += "Its master ID string seems to be [(!master_name || emagged) ? "empty" : master_name]."
+	//SKYRAT EDIT ADDITION BEGIN - CUSTOMIZATION
+	. += get_silicon_flavortext()
+	//SKYRAT EDIT ADDITION END
```
Comment (GoldenAlpharex) [inline code/modules/pai/pai.dm:175]: I'm not sure about this one, because that's starting to be a lot of different uses for silicon flavor-text.

Someone that plays as a borg often doesn't necessarily always want their borg examine text to be displayed on their pAI.

Something like this should be put in a different PR, possibly with its own flavor text pref.

File: code/modules/mob/mob.dm [hunk 12]
```diff
@@ -312,6 +323,16 @@
...
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
+	for(var/mob/camera/ai_eye/ai_eye in hearers)
+		if(ai_eye.ai.client && !(ai_eye.ai.stat == DEAD))
+			hearers -= ai_eye
+			hearers |= ai_eye.ai
+
+	for(var/obj/effect/overlay/holo_pad_hologram/holo in hearers)
+		if(holo.Impersonation?.client)
+			hearers |= holo.Impersonation
+	//SKYRAT EDIT ADDITION END - AI QoL
```
Comment (GoldenAlpharex) [inline code/modules/mob/mob.dm:336]:
```suggestion
	//SKYRAT EDIT ADDITION END - AI QoL

```

File: code/modules/mob/living/silicon/ai/examine.dm [hunk 13]
```diff
@@ -26,6 +26,9 @@
 			. += "The wireless networking light is blinking."
 		else if (!shunted && !client)
 			. += "[src]Core.exe has stopped responding! NTOS is searching for a solution to the problem..."
+	//SKYRAT EDIT ADDITION BEGIN - CUSTOMIZATION
+	. += get_silicon_flavortext()
+	//SKYRAT EDIT ADDITION END
```
Comment (GoldenAlpharex) [inline code/modules/mob/living/silicon/ai/examine.dm:31]: I'm not 100% on AI getting the general silicon flavortext, but I don't think it's a big deal in this specific case.

File: code/game/machinery/hologram.dm [hunk 14]
```diff
@@ -563,11 +541,12 @@ Possible to do for anyone motivated enough:
...
 			Hologram.Impersonation = user
 
-		Hologram.mouse_opacity = MOUSE_OPACITY_TRANSPARENT//So you can't click on it.
+		//Hologram.mouse_opacity = MOUSE_OPACITY_TRANSPARENT//So you can't click on it. //SKYRAT EDIT -- Customization; Making holograms clickable/examinable
 		Hologram.layer = FLY_LAYER //Above all the other objects/mobs. Or the vast majority of them.
 		SET_PLANE_EXPLICIT(Hologram, ABOVE_GAME_PLANE, src)
 		Hologram.set_anchored(TRUE)//So space wind cannot drag it.
-		Hologram.name = "[user.name] (Hologram)"//If someone decides to right click.
+		// Hologram.name = "[user.name] (Hologram)"//If someone decides to right click. //SKYRAT EDIT -- ORIGINAL
+		Hologram.name = user.name //SKYRAT EDIT -- Make the name exact, so that the double-emotes are less jarring in the chat
```
Comment (GoldenAlpharex) [inline code/game/machinery/hologram.dm:549]: I'm not sure that I understand this change. I think it makes sense that the hologram itself is marked as `(Hologram)`, what should happen is that the person sending the hologram should probably only see the hologram's own emote, and not the one coming from themselves, when they send it over hologram.

File: code/datums/emotes.dm [hunk 15]
```diff
@@ -122,12 +122,6 @@
 				continue
 			if(ghost.client.prefs.chat_toggles & CHAT_GHOSTSIGHT && !(ghost in viewers(user_turf, null)))
 				ghost.show_message("<span class='emote'>[FOLLOW_LINK(ghost, user)] [dchatmsg]</span>")
-	//SKYRAT EDIT ADDITION BEGIN - AI QoL
-	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
-		var/ai_eye_turf = get_turf(ai.eyeobj)
-		if(ai.client && !(ai.stat == DEAD) && (get_dist(user_turf, ai_eye_turf)<8))
-			ai.show_message("<span class='emote'>[dchatmsg]</span>")
-	//SKYRAT EDIT ADDITION END - AI QoL
```
Comment (GoldenAlpharex) [inline code/datums/emotes.dm:130]: So can AIs still see emotes now?

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/verbs/code/subtle.dm (hunk 6 above)
Comment (Useroth) [reply modular_skyrat/modules/verbs/code/subtle.dm]: Because despite being originally typed as mob, due to the subtler spaghetti code, the "target" var could hold a mob _or_ a number determining view distance that the emote is seen in.
With the addition of this PR, the "target" var might also be a hologram, which is also not a mob.

Also, originally, there's a show_message proc there that can only be used on mobs, so if I know it's a mob, i'm just casting the var to mob type.

File: modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm (hunk 9 above)
Comment (Useroth) [reply modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm:19]: As far as I can tell examining regular characters, we do.

File: modular_skyrat/master_files/code/game/machinery/hologram.dm (hunk 10 above)
Comment (Useroth) [reply modular_skyrat/master_files/code/game/machinery/hologram.dm]: It's a tg variable. One I probably should address some shit upstream.

File: code/modules/pai/pai.dm (hunk 11 above)
Comment (Useroth) [reply code/modules/pai/pai.dm:175]: This is done with assumption that people might use different characters for different applications.
Not gonna lie, you're not wrong that this should be its own thing. I'll probably do such later.

File: code/modules/mob/living/silicon/ai/examine.dm (hunk 13 above)
Comment (Useroth) [reply code/modules/mob/living/silicon/ai/examine.dm:31]: It could potentially be its own thing, like pAI, but currently it's so that the hologram can get nicely examined.

File: code/game/machinery/hologram.dm (hunk 14 above)
Comment (Useroth) [reply code/game/machinery/hologram.dm:549]: I did this for simplicity's sake, because doing otherwise is entering spaghetti territory and it's necessary to delve even deeper. Primarily because audible_message() doesn't have an argument where I can pass a list of ignored mobs, so two non-identical, due to different names, messages get doubleposted and clutter the chat.

File: code/datums/emotes.dm (hunk 15 above)
Comment (Useroth) [reply code/datums/emotes.dm:130]: Yes, handled in visible_message() and audible_message() respectively. I did it with the idea so that the AI can actually see the accompanying emotes' runechat text without doing too much snowflake code.

Though now that you mention it, I'll consider reworking it a bit later, because now I see a different way it could be done.

File: code/modules/mob/mob.dm (hunk 12 above)
Comment (Useroth) [reply code/modules/mob/mob.dm:336]: Done

File: modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm (hunk 3 above)
Comment (Useroth) [reply modular_skyrat/modules/customization/modules/mob/living/silicon/examine.dm:5]: Done

File: modular_skyrat/master_files/code/game/machinery/hologram.dm [hunk 16]
```diff
@@ -0,0 +1,14 @@
+GLOBAL_LIST_EMPTY(hologram_impersonators)
+
+/obj/machinery/holopad/set_holo(mob/living/user, obj/effect/overlay/holo_pad_hologram/h)
+	if(h.Impersonation)
+		GLOB.hologram_impersonators[user] = h
+		h.become_hearing_sensitive() // Well, we need to show up on "get_hearers_in_view()"
+	. = ..()
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/game/machinery/hologram.dm]: You forgot the `h`s here.

**Comment by Useroth** [comment]
(also the linter is screaming but I temporarily added a commit that reorders lines in the dme file, because otherwise it conflicted with another testmerged PR)

File: code/datums/emotes.dm [hunk 17]
```diff
@@ -137,6 +131,18 @@
...
+	if(hologram)
+		if(emote_type & (EMOTE_AUDIBLE | EMOTE_VISIBLE))
+			hologram.audible_message(msg, deaf_message = span_emote("You see how <b>[user]</b> [msg]"), audible_message_flags = EMOTE_MESSAGE)
+		else if(emote_type & EMOTE_VISIBLE)
+			hologram.visible_message(msg, visible_message_flags = EMOTE_MESSAGE)
+		if(emote_type & EMOTE_IMPORTANT)
+			for(var/mob/living/viewer in viewers(world.view, hologram))
+				if(viewer.is_blind() && !viewer.can_hear())
+					to_chat(viewer, msg)
+	// SKYRAT EDIT -- END
```
Comment (ghost) [inline code/datums/emotes.dm:145]: Think TG would want any of this?

Comment (Useroth) [reply code/datums/emotes.dm:145]: Possibly, but if I PRed it to tg, I'd probably need to rework hologram holos a bit better.
Also here, it's made so that stuff is relayed from the hologram itself for roleplay convenience.

Also they are in feature freeze too.

**Comment by ghost** [comment]
TG HAS shown interest in taking this up there by the way, so it might be worth leaving it TM until they take whatever they want

**Comment by JohnFulpWillard** [comment]
stop stalking our chat

**Review by GoldenAlpharex — APPROVED** [review]
Should be good to test-merge again.

**Review by GoldenAlpharex — APPROVED** [review]
